src/Cassandra/ProtocolEvents/SlidingWindowExclusiveTimer.cs

```diff
@@ -0,0 +1,182 @@
+﻿//
+//       Copyright (C) 2019 DataStax Inc.
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+//
+
+using System;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using Cassandra.Tasks;
+
+namespace Cassandra.ProtocolEvents
+{
+    /// <summary>
+    /// Timer that implements a sliding window style algorithm of delay which enables the timer to keep
+    /// postponing the action trigger while the <see cref="SlideDelay"/> keeps getting invoked (up to a maximum specified delay).
+    /// It also schedules all internal code execution with an exclusive scheduler and exposes it through <see cref="ExclusiveScheduler"/>.
+    /// </summary>
+    internal class SlidingWindowExclusiveTimer : IDisposable
+    {
+        private readonly ITimerFactory _timerFactory;
+        private readonly TimeSpan _delayIncrement;
+        private readonly TimeSpan _maxDelay;
+        private readonly Action _act;
+        private volatile ITimer _timer;
```

Comment:
I think we should use a single `readonly` timer instance, setting `dueTime` to infinite initially and call `Change()` when we want activate it.

Reply:
Unfortunately the `Change` method doesn't support the `state` parameter AFAIK and this implementation uses that parameter in order to avoid duplicate calls to `_act()`.

Reply:
Maybe I could experiment something like avoiding `System.Threading.Timer` completely and creating a `Task.Delay.ContinueWith(() => Fire(counter))`

Reply:
Do you think that the way it is currently will put too much GC pressure?

Reply:
I'm thinking about having multiple timers is more prone to race conditions, like timer is being disposed but at the same time a new timer is being created and those sort of stuff.

Reply:
all the code that creates, disposes, changes timers is running within the `Exclusive` scheduler so we should be good on that

Reply:
I just pushed an alternative implementation based on `Task.Delay`, check `TaskBasedTimer.cs`. Also, with this implementation, it's easy to change `Task.Delay` to our implementation of `HashedWheelTimer` if we want to. It supports passing a callback with `Change` and cancellation so that's all we need really.